File: src/diffbot_cpp_pkg/src/simple_parameter.cpp
#include <rclcpp/rclcpp.hpp>
#include <rcl_interfaces/msg/set_parameters_result.hpp>

class SimpleParameterNode : public rclcpp::Node
{
public:
    SimpleParameterNode() : Node("simple_parameter")
    {
        declare_parameter<int>("simple_int_param", 10);
        declare_parameter<std::string>("simple_str_param", "Diffbot Ros");

        param_callback_handle_ = add_on_set_parameters_callback(std::bind(&SimpleParameterNode::paramChangeCallback, this, std::placeholders::_1));
    }
private:
    OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;

    rcl_interfaces::msg::SetParametersResult paramChangeCallback(const std::vector<rclcpp::Parameter> & parameters)
    {
        rcl_interfaces::msg::SetParametersResult result;

        for(const auto& param: parameters)
        {
            if (param.get_name() == "simple_int_param" and param.get_type() == rclcpp::ParameterType::PARAMETER_INTEGER)
            {
                RCLCPP_INFO_STREAM(get_logger(), "Param simple_int_param have Changed! New Value is : " << param.as_int());
                result.successful = true;
            }

            if(param.get_name() == "simple_str_param" and param.get_type() == rclcpp::ParameterType::PARAMETER_STRING)
            {
                RCLCPP_INFO_STREAM(get_logger(), "Param simple_str_param have Changed! New Values is : " << param.as_string());
                result.successful = true;
            }
        }
        return result;
    }

};

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto simple_parameter = std::make_shared<SimpleParameterNode>();
    rclcpp::spin(simple_parameter);
    rclcpp::shutdown();
    return 0;
}

File: src/diffbot_py_pkg/diffbot_py_pkg/simple_parameter.py
import rclpy
from rclpy.node import Node
from rcl_interfaces.msg import SetParametersResult
from rclpy.parameter import Parameter

class SimpleParameter(Node):
    def __init__(self):
        super().__init__("simple_parameter")

        self.declare_parameter("simple_int_param", 10)
        self.declare_parameter("simple_str_param", "Diffbot Robot")

        self.add_on_set_parameters_callback(self.paramChangeCallback)
    
    def paramChangeCallback(self, params):
        result = SetParametersResult()
    
        for param in params:
            if param.name == "simple_int_param" and param.type_ == Parameter.Type.INTEGER:
                self.get_logger().info("Parame simple_int_param have Changed! New value is : %d" % param.value)
                result.successful = True
            
            if param.name == "simple_str_param" and param.type_ == Parameter.Type.STRING:
                self.get_logger().info("Param simple_str_param have Changed! New value is : %s" % param.value)
                result.successful = True

        return result
    
def main():
    rclpy.init()
    simple_parameter = SimpleParameter()
    rclpy.spin(simple_parameter)
    simple_parameter.destroy_node()
    rclpy.shutdown()



if __name__ == "__main__":
    main()
